crates/wasmtime/src/runtime/component/func/host.rs

```diff
@@ -372,61 +512,119 @@ where
     let func_ty = &types[ty];
     let param_tys = &types[func_ty.params];
     let result_tys = &types[func_ty.results];
-    let mut cx = LiftContext::new(store.0, &options, types, instance);
-    cx.enter_call();
-    if let Some(param_count) = param_tys.abi.flat_count(MAX_FLAT_PARAMS) {
-        // NB: can use `MaybeUninit::slice_assume_init_ref` when that's stable
-        let mut iter =
-            mem::transmute::<&[MaybeUninit<ValRaw>], &[ValRaw]>(&storage[..param_count]).iter();
-        args = param_tys
-            .types
-            .iter()
-            .map(|ty| Val::lift(&mut cx, *ty, &mut iter))
-            .collect::<Result<Box<[_]>>>()?;
-        ret_index = param_count;
-        assert!(iter.next().is_none());
+
+    let replay_enabled = store.0.replay_enabled();
+
+    rr_host_func_entry_event! { storage, &types[func_ty.params] => store.0 };
+
+    let results = if !replay_enabled {
```

Comment:
Same comment here as above re: dispatching at a higher level.

In general, the sense I'm getting is that this sequence evolved out of a desire to keep the overall shape the same, and add hooks where needed. But it seems to me that there are basically three possible flows at a high level:

- Normal host call;
- Host call, recorded;
- Replay of host call.

Could we instead write three different `lower_results` implementations depending on our mode, and dispatch once on the mode in `call_host`, to avoid the maze of fine-grained conditionals?